### PR TITLE
Generated files off in release

### LINKED ISSUE #5432: Turn off re-generation of source files in releases
In Mbed TLS 3.0, we decided to stop committing generated files in Git if they don't depend on the configuration or the platform. However, the generation of these files require tools that not everybody have (perl, python, a host C compiler, and starting with 3.2 some python modules to generate PSA driver wrapper code). So we had this idea that we'd include generated files in releases, so that users would not need to rebuild them.

We've updated our release process to commit the generated files. However, the build scripts still have dependencies on the source files, so depending on the timestamps of extracted files, they might end up getting rebuilt. ([Example](https://github.com/ARMmbed/mbedtls/issues/5374))

The goal of this issue is to completely turn off the generation of these files in a release. For CMake, the feature is present but not activated: we should toggle the default value of `GEN_FILES`. For Make, we need to create a similar option.

When this is done, update the [release process](https://confluence.arm.com/display/BSGSoftware/Shift+1) (Arm private link) accordingly.

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
* `make GEN_FILES=` now avoids rebuilding the configuration-independent generated files if not needed, similar (but not identical) to `cmake -DGEN_FILES=OFF`.
* In development, we default to re-generating configuration-independent files based on their dependencies. In releases, we bundle the files and avoid rebuilding them. Run `scripts/prepare_release.sh` to switch from development mode to release mode.

Resolves https://github.com/Mbed-TLS/mbedtls/issues/5432, in combination with adding “run `scripts/prepare_release.sh`” to the release preparation instructions.

Status: ready.

Fixes #5432 

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** not required
- [x] **backport** not required
- [ ] **tests** no (no new public interface); reviewers should test `make GEN_FILES=` manually

### CODE REVIEW

File: Makefile
Comment (tom-cosgrove-arm) [inline Makefile:54]: Doesn't this mean "If `GEN_FILES` is non-empty"? But line 48 sets `GEN_FILES` to `yes` if it's empty, so when is `GEN_FILES` going to be empty?

Comment (gilles-peskine-arm) [reply Makefile:54]: `GEN_FILES` is empty if you run `make GEN_FILES=` or after running `prepare_release.sh`.

File: scripts/prepare_release.sh
Comment (gilles-peskine-arm) [inline scripts/prepare_release.sh:10]: Do we really need an “unrelease” mode? I kept it from `gitignore_patch.sh`, but I'm not sure it's actually useful. We normally revert the release preparation commit(s) to go from release mode to development mode. Having this mode lets us do a bug fix on the release branch and get it back into development without merging, but we don't really need the “without merging” property?

File: programs/Makefile
Comment (tom-cosgrove-arm) [inline programs/Makefile:185]: Shouldn't this comment be in the root Makefile?

File: Makefile
Comment (tom-cosgrove-arm) [reply Makefile:54]: And if I run `make GEN_FILES=no` what happens?

Comment (gilles-peskine-arm) [reply Makefile:54]: Conditionals in GNU make are empty/undefined vs nonempty. `0`, `false`, `no`, 沒有``, etc. are true.

**Comment by tom-cosgrove-arm** [comment]
This looks okay to me. I might quibble with the name `prepare_release.sh`, but it's much better than `gitignore_patch.sh` (personally, since all it's doing is setting the "gen files" mode, I would prefer the filename to reflect that - I understand that other things could be added, but right now that seems to be an unnecessary extra level of indirection). But if you prefer `prepare_release.sh` that's fine.

The PR still has `needs-work` set - what else are you planning to do here?

**Comment by gilles-peskine-arm** [comment]
> I understand that other things could be added, but right now that seems to be an unnecessary extra level of indirection

We follow a script written in English for release preparation and I'd like to turn that into something the computer runs for us. So I want this script to do more in the medium term (not much more than this before the next release though): commit the generated files, assemble the changelog, possibly bump the version...

> The PR still has needs-work set - what else are you planning to do here?

I've been considering testing. This pull request doesn't add any public interface, so we don't need testing. But it would be nice to test `make GEN_FILES=` — and, preexisting, `cmake -DGEN_FILES=`. I'm looking into this right now and it's shaping up to be too complex to be included in this pull request, so I think I've decided this pull request won't include additional testing unless a reviewer requests it. Improved testing around generated files will come later through tech debt bandwidth.

**Comment by gilles-peskine-arm** [comment]
Also, I'm still considering dropping “unrelease” mode (in which case I'll rewrite the PR history). Any voices in favor of keeping it?

**Comment by tom-cosgrove-arm** [comment]
I like having the `unrelease` mode - it's not that much extra work (at least for now) and it allows the script to be symmetric

File: CMakeLists.txt
Comment (tom-cosgrove-arm) [inline CMakeLists.txt:60]: Should this be more emphatic? `significant` is doing a lot of work, here: specifically, `prepare_release.sh` will no longer work if the comment is changed too much

**Comment by gilles-peskine-arm** [comment]
> I like having the unrelease mode - it's not that much extra work

It's more than half the work of `prepare_release.sh`. Without it we wouldn't need option parsing, the mode variable and the magic comment in `CMakeLists.txt`.

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

File: CMakeLists.txt
Comment (minosgalanakis) [inline CMakeLists.txt:58]: Actively modifying a toggle option, using a hard to maintain sed inline replace string can be hard to maintain. 

I would separate the variable declaration and setting like that:

~~~~~~~
option(GEN_FILES "Generate the auto-generated files as needed" OFF)
if(CMAKE_HOST_WIN32)
    # N.B. The comment on the next line is significant! If you change it,
    # edit the sed command in prepare_release.sh that modifies
    # CMakeLists.txt.
    set(GEN_FILES OFF)
else()
    set(GEN_FILES ON)
endif()
~~~~~~~

Then replacing `set(GEN_FILES OFF)` -> `set(GEN_FILES ON)` can be easier. The only drawback to this is if we have checks that test if GEN_FILES  instead of its value (On=1, Off = 0) but I do not believe that is currently the the case. 

I won't block the PR for this minor issue though, so feel free to consider if only if you are making changes.

**Review by minosgalanakis — APPROVED** [review]
Looks good with a minor advisory

File: scripts/prepare_release.sh
Comment (minosgalanakis) [reply scripts/prepare_release.sh:10]: That would be outside of the scope of this PR/ASk though. I cannot speak for others since I have not used this particular script, but if people are using it it should stay. This is a helper tool for the maintainers anyway. We can always revisit this.

File: CMakeLists.txt
Comment (gilles-peskine-arm) [reply CMakeLists.txt:58]: > Actively modifying a toggle option, using a hard to maintain sed inline replace string can be hard to maintain.

I agree, but my preferred resolution would be to drop support for “unrelease” mode, which would remove most of the complexity around the sed expression.

Since neither you nor Tom consider this a blocker, I propose to merge this pull request and follow up later.

**Pull request merged**
